specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+
+### Search Tree
+
+The basic structure is that of an ordered search tree: The contained keys are organised such that they can be found (inserted, updated, ...) efficiently by value.
+
+*TODO Insert sketch of search tree*
+
+To efficiently find keys, the tree is traversed top-to-bottom and the non-leaf nodes help navigating/comparing the values efficiently. An intermediate node contains several ordered key-address pairs, which link to further nodes (intermediate or leaf) on the next lower level.
+
+Levels go from `0` representing Leaf Nodes, and go up for each level in the tree. The root of the tree will have the highest level in the tree and can give an estimate of it's overall size.
+
+Leaf nodes contain the actual Key-Value pairs for the tree which can be iterated over as part of the overall tree iteration.
+
+
+### Chunking
+
+Chunking is the strategy of determining chunk boundaries: Given a list of key-value pairs, it 'decides' which are still inside node A and which already go to the next node B on the same level.
+It depends on the hash/address/node of the items and a 'chunking factor' for tuning the shape of the tree.
+This chunking factor determines the average size of nodes/chunks that a node on a higher level contains.
+This average size in turn controls the shape of the tree (broad/narrow).
+The shape in turn defines the performance of operations on the tree.
+
+### Note on Multiple and Single Values and Sets
+
+The described tree can represent a data structure with multiple, single or no values per key. However, given that IPLD Maps (which a Prolly Tree loosely maps to) only allow one value for a key, implementations should merge duplicate keys into a single value. This is also important to have consistent ordering of key value pairs.
+
+### Re-Calculate
+
+In order to modify the tree (inserting, updating, or removing one or multiple values) a part of the tree needs to be re-calculated. This section describes an internal procedure that is not exposed to the caller.
+The tree is (partially) re-calculated bottom-up.
+
+Starting at the inserted/modified leaf or node referencing a now deleted/removed leaf, successively walk up the tree.If a removed item was the only item in its node: Remove the node from its parent and Check if the hash/address/CID of a new leaf/node splits the parent chunk (split the parent node if yes).
+    Then, check if a removed leaf/node was a 'splitting node' and the nodes need to be merged (only the last node in a chunk can be a splitting node). If splitting criterion holds for the last item of the node and there is the current node is succeeded by another node on the same level: merge the currend node with the succeeding node. Continue walking up the path. If at root and it is being split: Create a new root that links to the freshly split nodes. Return the new root CID.
+
+![](https://i.imgur.com/xE0id0V.png)
```

Comment:
Should we remove these pictures?

Reply:
Yeah sure if they aren't adding much to the spec anyway. 👍